language/stdlib/modules/libra_transaction_timeout.mvir

```diff
@@ -0,0 +1,47 @@
+module LibraTransactionTimeout {
+    import 0x0.LibraTimestamp;
+
+    resource Timeout {
+        // Only transactions with timestamp no later than block time + duration will be accepted.
+        duration: LibraTimestamp.Duration,
+    }
+
+    public initialize() {
+        let duration: LibraTimestamp.Duration;
+        let timeout: Self.Timeout;
+
+        // Only callable by the Association address
+        assert(get_txn_sender() == 0xA550C18, 1);
+
+        // Currently set to 10min.
+        duration = LibraTimestamp.new_duration(600);
+        timeout = Timeout {duration: move(duration)};
+        move_to_sender<Timeout>(move(timeout));
+        return;
+    }
+
+    public set_timeout(new_duration: LibraTimestamp.Duration) acquires Timeout {
+        let timeout: &mut Self.Timeout;
+
+        // Only callable by the Association address
+        assert(get_txn_sender() == 0xA550C18, 1);
+
+        timeout = borrow_global_mut<Timeout>(0xA550C18);
+        *(&mut move(timeout).duration) = move(new_duration);
+        return;
+    }
+
+    public is_valid_transaction_timestamp(timestamp: u64): bool acquires Timeout {
+        let current_block_time: LibraTimestamp.Instant;
+        let timeout: &Self.Timeout;
+        let duration: LibraTimestamp.Duration;
+        let txn_time: LibraTimestamp.Instant;
+
+        current_block_time = LibraTimestamp.get_global_time();
+        timeout = borrow_global<Timeout>(0xA550C18);
+        LibraTimestamp.add_duration(&mut current_block_time, *(&move(timeout).duration));
+
+        txn_time = LibraTimestamp.new_instant(move(timestamp));
+        return LibraTimestamp.is_before_exclusive(&current_block_time, &txn_time);
```

Comment:
This doesn't seem right to me. This procedure should check two things:

(1) [freshness] `timestamp` is greater than the current time
(2) [not too far in the future] timestamp is less than `current_block_time + timeout`

This seems to be checking `current_block_time + duration < timestamp` instead.

Reply:
Also, I think we need a bunch of tests for this timeout feature. Currently, there are none.

It might be easier to wait on actually adding the timestamp expiration feature to the prologue in a future PR that we can scrutinize more carefully. There's already a lot going on here.